File: 프로그래머스/2/150369. 택배 배달과 수거하기/택배 배달과 수거하기.py
# cap = 최대 적재량
# n = 집 수, deliveries, pickups의 최대길이와 일치
# deliveries -> 배달할 택배 배열
# pickups -> 수거할 상자 배열
def solution(cap, n, deliveries, pickups):
    answer = 0
    deliver = 0
    pickup = 0
    
    # 가장 먼 집부터 해결하기
    
    for i in range(n-1, -1, -1):
        deliver += deliveries[i]
        pickup += pickups[i]
        
        cnt = 0
        while deliver > 0 or pickup > 0:
            deliver -= cap
            pickup -= cap
            cnt += 1
        
        if cnt > 0:
            answer += (i + 1) * 2 * cnt
    
    return answer

File: 프로그래머스/2/150369. 택배 배달과 수거하기/택배 배달과 수거하기.java
class Solution {
    // cap : 최대 적재량
    // n 집 수 -> deliveries.len, pickups.len
    // deliveries : 배달 대상 배열
    // pickups : 수거할 택배 상자 배열
    public long solution(int cap, int n, int[] deliveries, int[] pickups) {
        long answer = 0;
        int deliver = 0;
        int pickup = 0;
        
        // 가장 먼 집부터 처리
        for (int i = n - 1; i >= 0; i--) {
            deliver += deliveries[i];
            pickup += pickups[i];
            
            // 현재 위치까지 왕복해야 하는 횟수 계산
            int cnt = 0;
            while (deliver > 0 || pickup > 0) {
                deliver -= cap;
                pickup -= cap;
                cnt++;
            }
            
            // 왕복 거리 계산
            if (cnt > 0) {
                answer += (i + 1) * 2 * cnt;
            }
        }
        
        return answer;
    }
}
